Caching in Mantis Master API v1 for listing jobs.

codyrioux opened the pull request:
### Context
The jobs list call is expensive having to reach out to all actors in the system and wait for a response. Some short term caching as discussed in the team meeting will greatly reduce the frequency with which the API needs to wait for a response from all actors.

### Checklist

- [x] `./gradlew build` compiles code correctly
- [x] Added new tests where applicable
- [x] `./gradlew test` passes all tests
- [x] Extended README or added javadocs where applicable
- [x] Added copyright headers for new files from `CONTRIBUTING.md`

codyrioux committed:
DRAFT: Caching in Mantis Master API v1 for listing jobs.

codyrioux commented on server/src/main/java/io/mantisrx/master/api/akka/route/v1/JobsRoute.java:
```diff
@@ -228,9 +242,17 @@ private Route getJobsRoute(Optional<String> clusterName) {
+
+                                        if (config.getApiCacheEnabled() && cachedResponse == null) {
+                                            listJobsCache.put(listJobsRequest, resp);
+                                        }
```
I chose this approach instead of the `Cache#get(key, supplier)` approach for two reasons:

1. I wasn't sure if the `CompletionStage<>` from `jobRouteHandler#listJobs` would become a `CompletedFuture<>` or equivalent afterwards or become unusable. Otherwise a cache of `CompletableFuture` would be very convenient.

2. The `Cache#get` method throws a `java.util.concurrent.ExecutionException` and I was of the opinion that this was less ugly than dealing with the exception.

nickmahilani commented:
we have been using the akka api caching (backed by Caffeine), this can be controlled with cache control headers if we chose to do that later on.. Example https://github.com/Netflix/mantis-control-plane/blob/master/server/src/main/java/io/mantisrx/master/api/akka/route/v1/JobClustersRoute.java#L192

codyrioux commented:
Thanks Nick! This looks as though it integrates much more cleanly and responding to cache control headers is always a good option to keep open in the future. I'll get this PR updated.

codyrioux committed:
Standardized caching configuration, JobsRoute caches.

Routes on v0 and v1 now all respect the new configuration options added
to MasterConfiguration. Specifically the min/initial and max cache size
as well as TTL (milliseconds) for the caches.

Added caching to v1 JobsRoute which did not previously perform any
caching.

codyrioux commented:
Leaned into the existing caching framework as well as standardized the cache configurations around some newly added configuration options in the `MasterConfiguration` class.

codyrioux commented:
Merged via rebase/merge in git.
